xmtp_mls/src/groups/device_sync.rs

```diff
@@ -0,0 +1,596 @@
+    V: SmartContractSignatureVerifier + Clone,
+{
+    pub async fn enable_history_sync(&self) -> Result<(), GroupError> {
+        // look for the sync group, create if not found
+        let sync_group = match self.get_sync_group() {
+            Ok(group) => group,
+            Err(_) => {
+                // create the sync group
+                self.create_sync_group()?
+            }
+        };
+
+        // sync the group
+        sync_group.sync().await?;
+
+        Ok(())
+    }
+
+    async fn send_sync_request(
+        &self,
+        request: DeviceSyncRequest,
+    ) -> Result<(String, String), DeviceSyncError> {
+        // find the sync group
+        let conn = self.store().conn()?;
+        let sync_group = self.get_sync_group()?;
+
+        // sync the group
+        sync_group.sync().await?;
+
+        // lookup if a request has already been made
+        if let Ok((_msg, request)) = self.pending_sync_request(request.kind).await {
+            return Ok((request.request_id, request.pin_code));
+        }
+
+        // build the request
+        let request: DeviceSyncRequestProto = request.into();
+        let pin_code = request.pin_code.clone();
+        let request_id = request.request_id.clone();
+
+        let content = DeviceSyncContent::Request(request.clone());
+        let content_bytes = serde_json::to_vec(&content)?;
+
+        let _message_id = sync_group.prepare_message(&content_bytes, &conn, move |_time_ns| {
+            PlaintextEnvelope {
+                content: Some(Content::V2(V2 {
+                    message_type: Some(Request(request)),
+                    idempotency_key: new_request_id(),
+                })),
+            }
+        })?;
+
+        // publish the intent
+        if let Err(err) = sync_group.publish_intents(&conn.into()).await {
+            tracing::error!("error publishing sync group intents: {:?}", err);
+        }
+
+        Ok((request_id, pin_code))
+    }
+
+    async fn send_sync_reply(&self, contents: DeviceSyncReplyProto) -> Result<(), DeviceSyncError> {
+        // find the sync group
+        let conn = self.store().conn()?;
+        let sync_group = self.get_sync_group()?;
+
+        // sync the group
+        sync_group.sync().await?;
+
+        let (msg, _request) = self.pending_sync_request(contents.kind()).await?;
+
+        // add original sender to all groups on this device on the node
+        self.ensure_member_of_all_groups(&msg.sender_inbox_id)
+            .await?;
+
+        // the reply message
+        let (content_bytes, contents) = {
+            let content = DeviceSyncContent::Reply(contents);
+            let content_bytes = serde_json::to_vec(&content)?;
+            let DeviceSyncContent::Reply(contents) = content else {
+                unreachable!("This is a reply.");
+            };
+
+            (content_bytes, contents)
+        };
+
+        let _message_id = sync_group.prepare_message(&content_bytes, &conn, move |_time_ns| {
```

Comment:
noob Q: How come we need both `content_bytes` and `content` here? aren't they the same thing?

Reply:
Not sure. This is legacy code that I moved around, but didn't really check. Let me do some poking around to see what I can find.

Reply:
Looking at this a bit more, we're not using both. The `content_bytes` is serialized, and we're not actually using the `content` value here, that's just a field. :) Maybe you looked over it too quickly?